Question 1
Refer to these Stata DTA file: Countrie_ dependencie_ area. This is a list of the world’s countries
and their dependencies by land, water, and total area. For more details, visit the Wikipedia page
• Calculate the percentage of water in relation to the total land area measured in kilomete 100 (1)
• Calculate the percentage of water in relation to the total land area measured 

× 100 (2)
• Find out which 10 countries have the largest land (in km2) with water in the world.
• Find out which 10 countries have the water (%) in the world.

In [156]:
import pandas as pd
import numpy as np
import re
file_path = 'C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Countries_dependencies_area.dta'
data = pd.read_stata(file_path)
df = pd.DataFrame(data)
print(df.head())
print(df.tail())
df['Country_Name'].nunique()

  Country_Name           Area_Type            Area_Size ISO_Alpha3_Code
0  Afghanistan  Total in km2 (mi2)    652,864 (252,072)             AFG
1  Afghanistan   Land in km2 (mi2)    652,230 (251,830)             AFG
2      Albania  Total in km2 (mi2)      28,748 (11,100)             ALB
3      Albania   Land in km2 (mi2)      27,400 (10,600)             ALB
4      Algeria  Total in km2 (mi2)  2,381,741 (919,595)             DZA
    Country_Name           Area_Type          Area_Size ISO_Alpha3_Code
381        Yemen   Land in km2 (mi2)  527,968 (203,850)             YEM
382       Zambia   Land in km2 (mi2)  743,390 (287,020)             ZMB
383       Zambia  Total in km2 (mi2)  752,612 (290,585)             ZMB
384     Zimbabwe  Total in km2 (mi2)  390,757 (150,872)             ZWE
385     Zimbabwe   Land in km2 (mi2)  386,850 (149,360)             ZWE


193

In [108]:
#Function to separate km² and mi² from Area_Size
def separate_area_sizes(df):
    df[['Area_km2', 'Area_mi2']] = df['Area_Size'].str.extract(r'([\d,]+) \(([\d,]+)\)')
    
    # Remove commas and convert to numeric
    df['Area_km2'] = df['Area_km2'].str.replace(',', '').astype(float)
    df['Area_mi2'] = df['Area_mi2'].str.replace(',', '').astype(float)

    return df

# Apply the function
df = separate_area_sizes(df)

# Pivot the DataFrame to create separate columns for Total and Land Areas
area_df = df.pivot_table(index='Country_Name', 
                          columns='Area_Type', 
                          values=['Area_km2', 'Area_mi2'], 
                          aggfunc='first')

# Reset index to make Country_Name a column again
area_df.reset_index(inplace=True)

# Rename columns for clarity
area_df.columns = ['Country_Name', 
                   'Land_Area_km2', 'Total_Area_km2', 
                   'Land_Area_mi2', 'Total_Area_mi2']

# Display the resulting DataFrame
print(area_df)

    Country_Name  Land_Area_km2  Total_Area_km2  Land_Area_mi2  Total_Area_mi2
0    Afghanistan       652230.0        652864.0       251830.0        252072.0
1        Albania        27400.0         28748.0        10600.0         11100.0
2        Algeria      2381741.0       2381741.0       919595.0        919595.0
3        Andorra          468.0           468.0          181.0           181.0
4         Angola      1246700.0       1246700.0       481350.0        481350.0
..           ...            ...             ...            ...             ...
187    Venezuela       882050.0        912050.0       340560.0        352140.0
188      Vietnam       313429.0        331340.0       121016.0        127930.0
189        Yemen       527968.0        527968.0       203850.0        203850.0
190       Zambia       743390.0        752612.0       287020.0        290585.0
191     Zimbabwe       386850.0        390757.0       149360.0        150872.0

[192 rows x 5 columns]


In [110]:
#Check the data type 
print(df.dtypes)

Country_Name        object
Area_Type           object
Area_Size           object
ISO_Alpha3_Code     object
Area_km2           float64
Area_mi2           float64
dtype: object


In [124]:
#Question 1: Calculate Water Percentage (in km2) (Euqation1)
area_df['Water_Percentage_km2'] = ((area_df['Total_Area_km2'] - area_df['Land_Area_km2']) / area_df['Total_Area_km2']) * 100

# Display the resulting DataFrame with Water Percentage
print(area_df[['Country_Name', 'Total_Area_km2', 'Land_Area_km2', 'Water_Percentage_km2']])

    Country_Name  Total_Area_km2  Land_Area_km2  Water_Percentage_km2
0    Afghanistan        652864.0       652230.0              0.097111
1        Albania         28748.0        27400.0              4.689022
2        Algeria       2381741.0      2381741.0              0.000000
3        Andorra           468.0          468.0              0.000000
4         Angola       1246700.0      1246700.0              0.000000
..           ...             ...            ...                   ...
187    Venezuela        912050.0       882050.0              3.289293
188      Vietnam        331340.0       313429.0              5.405626
189        Yemen        527968.0       527968.0              0.000000
190       Zambia        752612.0       743390.0              1.225333
191     Zimbabwe        390757.0       386850.0              0.999854

[192 rows x 4 columns]


In [114]:
#To solve Equation 2
# Calculate Water Percentage in square miles
area_df['Water_Percentage_mi2'] = ((area_df['Total_Area_mi2'] - area_df['Land_Area_mi2']) / area_df['Total_Area_mi2']) * 100

# Display the resulting DataFrame with Water Percentage in mi²
print(area_df[['Country_Name', 'Total_Area_mi2', 'Land_Area_mi2', 'Water_Percentage_mi2']])

    Country_Name  Total_Area_mi2  Land_Area_mi2  Water_Percentage_mi2
0    Afghanistan        252072.0       251830.0              0.096004
1        Albania         11100.0        10600.0              4.504505
2        Algeria        919595.0       919595.0              0.000000
3        Andorra           181.0          181.0              0.000000
4         Angola        481350.0       481350.0              0.000000
..           ...             ...            ...                   ...
187    Venezuela        352140.0       340560.0              3.288465
188      Vietnam        127930.0       121016.0              5.404518
189        Yemen        203850.0       203850.0              0.000000
190       Zambia        290585.0       287020.0              1.226836
191     Zimbabwe        150872.0       149360.0              1.002174

[192 rows x 4 columns]


In [122]:
#Question 1: Find out which 10 countries have the largest land (in km2) with water in the world.
# Calculate Water Area
area_df['Water_Area_km2'] = area_df['Total_Area_km2'] - area_df['Land_Area_km2']

# Sort by Land Area and select the top 10
top_10_largest_land = area_df.nlargest(10, 'Land_Area_km2')

# Display the resulting DataFrame with Land and Water Areas
print(top_10_largest_land[['Country_Name', 'Land_Area_km2', 'Water_Area_km2']])

                   Country_Name  Land_Area_km2  Water_Area_km2
140                      Russia     16376870.0        721376.0
132  People's Republic of China      9326410.0        270550.0
183               United States      9147593.0        377474.0
31                       Canada      9093507.0        891163.0
23                       Brazil      8460415.0         49931.0
8                     Australia      7682300.0         58920.0
74                        India      2973190.0        314073.0
6                     Argentina      2736690.0         43710.0
85                   Kazakhstan      2699700.0         25210.0
2                       Algeria      2381741.0             0.0


In [126]:
#Question 1: Find out which 10 countries have the water (%) in the world.
# Sort by Water Percentage and select the top 10
top_10_highest_water_percentage = area_df.nlargest(10, 'Water_Percentage_km2')

# Display the resulting DataFrame with Country Name and Water Percentage
print(top_10_highest_water_percentage[['Country_Name', 'Water_Percentage_km2']])

      Country_Name  Water_Percentage_km2
11         Bahamas             27.881844
68   Guinea-Bissau             22.159170
100         Malawi             20.596874
118    Netherlands             19.042159
179         Uganda             16.986131
51         Eritrea             14.115646
94         Liberia             13.512737
13      Bangladesh             12.319817
60          Gambia             10.402833
57         Finland             10.152449


Question 2: Refer to the ISO 3166 Countries Codes CSV file. This file contains official state names, sovereignty
status—indicating if a country is UN member, disputed, or under another state’s control—and TLD (Top-
Level Domain), which is the Internet domain suffix for each country (e.g., .pk for Pakistan). For more
details, visit the Wikipedia page.
• Read the CSV and assign the Name sovereign UN memb df to this Dataframe.
• Clean the Data: Remove any unnecessary notes, references, or sources attached to the values in
any column. For example remove the [e] from Antarctica[e] or [c][d] from ˚Aland[c][d]
• Filter UN Member Countries: Eliminate any countries from the dataset that are not UN
members. You can determine UN membership by checking the “Sovereignty Status” column.
• Reset index Remember, there are currently 193 sovereign states that are members of the United
Nations.Ensure that the index is reset properly.

In [142]:
# Read the CSV file
countries_data = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/ISO_3166_Countries_Codes.csv')
sovereign_UN_memb_df = pd.DataFrame(sovereign_UN_memb_df)
print(sovereign_UN_memb_df.head())
print(sovereign_UN_memb_df.tail())

    ISO_3166_Name                          Official_State_Name  \
0     Afghanistan       the Islamic Republic of Afghanistan[b]   
1   Åland Islands                                  Åland[c][d]   
2         Albania                      the Republic of Albania   
3         Algeria  the People's Democratic Republic of Algeria   
4  American Samoa                            American Samoa[c]   

  Sovereignty_Status ISO_Alpha3_Code Top_Level_Domain  
0          UN member             AFG              .af  
1            Finland             ALA              .ax  
2          UN member             ALB              .al  
3          UN member             DZA              .dz  
4      United States             ASM              .as  
           ISO_3166_Name                             Official_State_Name  \
266    Wallis and Futuna  the Territory of the Wallis and Futuna Islands   
267  Western Sahara [an]      the Sahrawi Arab Democratic Republic[c][d]   
268                Yemen               

In [162]:
# Function to clean the values in a column
def clean_value(value):
    # Remove references in square brackets
    return re.sub(r'\s*\[.*?\]', '', value)

# Apply the cleaning function to all columns in the DataFrame
for column in sovereign_UN_memb_df.columns:
    sovereign_UN_memb_df[column] = sovereign_UN_memb_df[column].map(clean_value)

# Display the cleaned DataFrame
print(sovereign_UN_memb_df.head())

    ISO_3166_Name                          Official_State_Name  \
0     Afghanistan          the Islamic Republic of Afghanistan   
1   Åland Islands                                        Åland   
2         Albania                      the Republic of Albania   
3         Algeria  the People's Democratic Republic of Algeria   
4  American Samoa                               American Samoa   

  Sovereignty_Status ISO_Alpha3_Code Top_Level_Domain  
0          UN member             AFG              .af  
1            Finland             ALA              .ax  
2          UN member             ALB              .al  
3          UN member             DZA              .dz  
4      United States             ASM              .as  


In [182]:
# Filter the DataFrame for UN member countries
sovereign_UN_memb_df = sovereign_UN_memb_df[sovereign_UN_memb_df["Sovereignty_Status"] == "UN member"]

# Display the filtered DataFrame
print(sovereign_UN_memb_df.head())


  ISO_3166_Name                          Official_State_Name  \
0   Afghanistan          the Islamic Republic of Afghanistan   
2       Albania                      the Republic of Albania   
3       Algeria  the People's Democratic Republic of Algeria   
5       Andorra                  the Principality of Andorra   
6        Angola                       the Republic of Angola   

  Sovereignty_Status ISO_Alpha3_Code Top_Level_Domain  
0          UN member             AFG              .af  
2          UN member             ALB              .al  
3          UN member             DZA              .dz  
5          UN member             AND              .ad  
6          UN member             AGO              .ao  


In [184]:
# Reset the index of the filtered DataFrame
sovereign_UN_memb_df.reset_index(drop=True, inplace=True)

# Display the DataFrame with the reset index
print(sovereign_UN_memb_df.head())
print(sovereign_UN_memb_df.tail())

# Print the number of countries to confirm it matches the expected count
print(f"Total number of UN member countries: {len(un_member_df)}")

  ISO_3166_Name                          Official_State_Name  \
0   Afghanistan          the Islamic Republic of Afghanistan   
1       Albania                      the Republic of Albania   
2       Algeria  the People's Democratic Republic of Algeria   
3       Andorra                  the Principality of Andorra   
4        Angola                       the Republic of Angola   

  Sovereignty_Status ISO_Alpha3_Code Top_Level_Domain  
0          UN member             AFG              .af  
1          UN member             ALB              .al  
2          UN member             DZA              .dz  
3          UN member             AND              .ad  
4          UN member             AGO              .ao  
                          ISO_3166_Name                   Official_State_Name  \
188  Venezuela (Bolivarian Republic of)  the Bolivarian Republic of Venezuela   
189                            Viet Nam    the Socialist Republic of Viet Nam   
190                               Ye

Question 3
Please refer to the world bank country groups CSV file. This file contains the The World Bank
Group classification of the world’s economies, classified by income, region, and World Bank lending
status. Visit the interactive version on the World Bank Country Groups
• Read the CSV and assign the Name wb groups df to this Dataframe.
• Merge Datasets: Combine wb groups df with the sovereign UN memb df.
• Name the merged dataframe as countries df.
• Include “merge indicator” column to show the status of each row from both datasets.
• Consider using the column ISO Alpha3 Code while merging the data frames.
• Verify that all 193 UN member countries are matched by cross-tabulating the “merge indicator”
and “sovereignty status” columns.
• Drop and Reset: Keep only the rows that are present in both datasets, resulting in a total of 193
rows. Drop the “merge indicator” and “Code” columns. Ensure that the index is reset properly.

In [180]:
#Part(a): Read the CSV and assign the Name wb groups df to this Dataframe.
# Read the CSV file
wb_groups_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/world_bank_groups.csv')

# Display the first few rows of the DataFrame to understand its structure
print(wb_groups_df.head())

          Economy Code                      Region         Income group  \
0     Afghanistan  AFG                  South Asia           Low income   
1         Albania  ALB       Europe & Central Asia  Upper middle income   
2         Algeria  DZA  Middle East & North Africa  Upper middle income   
3  American Samoa  ASM         East Asia & Pacific          High income   
4         Andorra  AND       Europe & Central Asia          High income   

  Lending category  
0              IDA  
1             IBRD  
2             IBRD  
3              NaN  
4              NaN  


In [194]:
#Part (b): Merge datasets: Combine wb groups df with the sovereign UN memb df.
# Merge the DataFrames on the 'ISO_Alpha3_Code' and 'Code'
countries_df = pd.merge(
    sovereign_UN_memb_df,
    wb_groups_df,
    left_on="ISO_Alpha3_Code",
    right_on="Code",
    how="inner",  
    indicator=True  
)

# Display the merged DataFrame
print(countries_df.head())

print(f"Total number of UN member countries: {len(countries_df)}")

  ISO_3166_Name                          Official_State_Name  \
0   Afghanistan          the Islamic Republic of Afghanistan   
1       Albania                      the Republic of Albania   
2       Algeria  the People's Democratic Republic of Algeria   
3       Andorra                  the Principality of Andorra   
4        Angola                       the Republic of Angola   

  Sovereignty_Status ISO_Alpha3_Code Top_Level_Domain      Economy Code  \
0          UN member             AFG              .af  Afghanistan  AFG   
1          UN member             ALB              .al      Albania  ALB   
2          UN member             DZA              .dz      Algeria  DZA   
3          UN member             AND              .ad      Andorra  AND   
4          UN member             AGO              .ao       Angola  AGO   

                       Region         Income group Lending category _merge  
0                  South Asia           Low income              IDA   both  
1       Eu

In [196]:
# Cross-tabulate the “merge indicator” and “Sovereignty_Status”
cross_tab = pd.crosstab(countries_df['_merge'], countries_df['Sovereignty_Status'])

print(cross_tab)

Sovereignty_Status  UN member
_merge                       
both                      193


In [198]:
# Drop the "merge indicator" and "Code" columns
countries_df = countries_df.drop(columns=['_merge', 'Code'])

# Reset the index
countries_df.reset_index(drop=True, inplace=True)

# Display the final DataFrame
print(countries_df)
print(f"Total number of rows: {len(countries_df)}")

                          ISO_3166_Name  \
0                           Afghanistan   
1                               Albania   
2                               Algeria   
3                               Andorra   
4                                Angola   
..                                  ...   
188  Venezuela (Bolivarian Republic of)   
189                            Viet Nam   
190                               Yemen   
191                              Zambia   
192                            Zimbabwe   

                             Official_State_Name Sovereignty_Status  \
0            the Islamic Republic of Afghanistan          UN member   
1                        the Republic of Albania          UN member   
2    the People's Democratic Republic of Algeria          UN member   
3                    the Principality of Andorra          UN member   
4                         the Republic of Angola          UN member   
..                                           ...         

Question 4
Please refer to the internet speedtest global index CSV file. The Speedtest Global Index includes
internet speed Ranking of mobile and fixed broadband speeds from around the world. For country wise
Speedtest Intelligence, you can visit the Speedtest Global Index by Ookla.
• Read the CSV and assign the Name ookla df to this Dataframe.
• Merge Datasets: Combine ookla df with the countries df.
• Name the merged dataframe as speedtest index df.
• Consider using the column ISO Alpha3 Code while merging the data frames.
• Find out which countries have the fastest, and slowest internet speeds in the world the in terms of
Fixed Broadband speed
• Find out which countries have the fastest, and slowest internet speeds in the world the in terms of
Mobile speed
• Compute the minimum and maximum Fixed Broadband speed Mbps for each 1) region, 2) income
group?
• Compute the minimum and maximum Mobile Speed Mbps for each 1) region, 2) income group?
• Print top 25 countries based on Fixed Broadband Mbps.
• Print top 25 countries based on Mobile Speed Mbps.
• Filter the region as “South Asia“, Print three columns country, Mobile , and Fixed Broadband
Speeds
• Additional optional interactive resources to explore: Cloud Infrastructure Map around the World,
Submarine Cable Map, and more. And of course, let’s try not to get too distracted by where the
sharks are dining on Pakistan’s ocean cables!
• Can you write a few lines about the state of the internet, technology, and data? Reflect on where
we stand in this relentless race.

In [200]:
# Read the CSV file
ookla_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/internet speedtest global index.csv')

# Display the first few rows of the DataFrame to understand its structure
print(ookla_df.head())

  ISO 3166-1 alpha-3  Fixed Broadband Mbps  Mobile Speed Mbps
0                AFG                  3.63               8.48
1                ALB                 62.71              50.82
2                DZA                 14.26              24.90
3                AND                   NaN                NaN
4                AGO                 18.58                NaN


In [202]:
# Merge Datasets: Combine ookla_df with the countries_df.
# Name the merged dataframe as speedtest_index_df.
# Consider using the column ISO_Alpha3_Code while merging the data frames.

speedtest_index_df = pd.merge(
    countries_df,
    ookla_df,
    left_on="ISO_Alpha3_Code",
    right_on="ISO 3166-1 alpha-3",
    how="outer"  
)

# Display the merged DataFrame
print(speedtest_index_df)

                  ISO_3166_Name                  Official_State_Name  \
0                   Afghanistan  the Islamic Republic of Afghanistan   
1                        Angola               the Republic of Angola   
2                       Albania              the Republic of Albania   
3                       Andorra          the Principality of Andorra   
4    United Arab Emirates (the)             the United Arab Emirates   
..                          ...                                  ...   
188                       Samoa       the Independent State of Samoa   
189                       Yemen                the Republic of Yemen   
190                South Africa         the Republic of South Africa   
191                      Zambia               the Republic of Zambia   
192                    Zimbabwe             the Republic of Zimbabwe   

    Sovereignty_Status ISO_Alpha3_Code Top_Level_Domain               Economy  \
0            UN member             AFG              .a

In [204]:

# Find the index of the country with the fastest fixed broadband speed
fastest_index = speedtest_index_df['Fixed Broadband Mbps'].idxmax()
fastest_country = speedtest_index_df.loc[fastest_index]

# Find the index of the country with the slowest fixed broadband speed
slowest_index = speedtest_index_df['Fixed Broadband Mbps'].idxmin()
slowest_country = speedtest_index_df.loc[slowest_index]

# Display the results
print("Fastest Fixed Broadband Speed:")
print(f"Country Code: {fastest_country['ISO 3166-1 alpha-3']}, Speed: {fastest_country['Fixed Broadband Mbps']} Mbps")

print("\nSlowest Fixed Broadband Speed:")
print(f"Country Code: {slowest_country['ISO 3166-1 alpha-3']}, Speed: {slowest_country['Fixed Broadband Mbps']} Mbps")

Fastest Fixed Broadband Speed:
Country Code: ARE, Speed: 297.62 Mbps

Slowest Fixed Broadband Speed:
Country Code: TKM, Speed: 3.53 Mbps


In [206]:
# To find out fastest and slowest Mobile Speed
# Drop rows with NaN in 'Mobile Speed Mbps'
speedtest_index_df_cleaned = speedtest_index_df.dropna(subset=['Mobile Speed Mbps'])

# Find the index of the country with the fastest mobile speed
fastest_mobile_index = speedtest_index_df_cleaned['Mobile Speed Mbps'].idxmax()
fastest_mobile_country = speedtest_index_df_cleaned.loc[fastest_mobile_index]

# Find the index of the country with the slowest mobile speed
slowest_mobile_index = speedtest_index_df_cleaned['Mobile Speed Mbps'].idxmin()
slowest_mobile_country = speedtest_index_df_cleaned.loc[slowest_mobile_index]

# Display the results
print("Fastest Mobile Speed:")
print(f"Country Code: {fastest_mobile_country['ISO 3166-1 alpha-3']}, Speed: {fastest_mobile_country['Mobile Speed Mbps']} Mbps")

print("\nSlowest Mobile Speed:")
print(f"Country Code: {slowest_mobile_country['ISO 3166-1 alpha-3']}, Speed: {slowest_mobile_country['Mobile Speed Mbps']} Mbps")

Fastest Mobile Speed:
Country Code: ARE, Speed: 398.51 Mbps

Slowest Mobile Speed:
Country Code: YEM, Speed: 7.91 Mbps


In [212]:
# Compute minimum and maximum Fixed Broadband speed for each region
region_stats = speedtest_index_df.groupby('Region')['Fixed Broadband Mbps'].agg(['min', 'max']).reset_index()
print("Minimum and Maximum Fixed Broadband Speed by Region:")
print(region_stats)

# Compute minimum and maximum Fixed Broadband speed for each income group
income_group_stats = speedtest_index_df.groupby('Income group')['Fixed Broadband Mbps'].agg(['min', 'max']).reset_index()
print("\nMinimum and Maximum Fixed Broadband Speed by Income Group:")
print(income_group_stats)

Minimum and Maximum Fixed Broadband Speed by Region:
                       Region     min     max
0         East Asia & Pacific   21.56  297.57
1       Europe & Central Asia    3.53  223.72
2   Latin America & Caribbean   12.09  265.62
3  Middle East & North Africa    3.65  297.62
4               North America  188.20  242.27
5                  South Asia    3.63   71.62
6          Sub-Saharan Africa    8.53   62.04

Minimum and Maximum Fixed Broadband Speed by Income Group:
          Income group    min     max
0          High income  40.58  297.62
1           Low income   3.63   45.80
2  Lower middle income   9.35  151.69
3  Upper middle income   3.53  232.15


In [214]:
# Compute minimum and maximum Mobile Speed for each region
region_mobile_stats = speedtest_index_df.groupby('Region')['Mobile Speed Mbps'].agg(['min', 'max']).reset_index()
print("Minimum and Maximum Mobile Speed by Region:")
print(region_mobile_stats)

# Compute minimum and maximum Mobile Speed for each income group
income_group_mobile_stats = speedtest_index_df.groupby('Income group')['Mobile Speed Mbps'].agg(['min', 'max']).reset_index()
print("\nMinimum and Maximum Mobile Speed by Income Group:")
print(income_group_mobile_stats)

Minimum and Maximum Mobile Speed by Region:
                       Region    min     max
0         East Asia & Pacific  28.70  141.23
1       Europe & Central Asia  11.98  133.44
2   Latin America & Caribbean  11.26   73.08
3  Middle East & North Africa   7.91  398.51
4               North America  74.87  103.26
5                  South Asia   8.48   96.38
6          Sub-Saharan Africa  20.19   51.96

Minimum and Maximum Mobile Speed by Income Group:
          Income group    min     max
0          High income  21.79  398.51
1           Low income   7.91   12.79
2  Lower middle income  11.26   96.38
3  Upper middle income  11.98  108.37


In [216]:
# Sort the DataFrame by 'Fixed Broadband Mbps' in descending order and select the top 25
top_25_fixed_broadband = speedtest_index_df.sort_values(by='Fixed Broadband Mbps', ascending=False).head(25)

# Print the top 25 countries
print("Top 25 Countries Based on Fixed Broadband Speed (Mbps):")
print(top_25_fixed_broadband[['ISO 3166-1 alpha-3', 'Fixed Broadband Mbps']])

Top 25 Countries Based on Fixed Broadband Speed (Mbps):
    ISO 3166-1 alpha-3  Fixed Broadband Mbps
4                  ARE                297.62
151                SGP                297.57
31                 CHL                265.62
182                USA                242.27
169                THA                232.15
58                 FRA                223.72
47                 DNK                219.23
81                 ISL                213.98
82                 ISR                206.42
53                 ESP                205.17
30                 CHE                203.50
145                ROU                200.94
86                 JPN                197.78
32                 CHN                189.49
94                 KWT                188.29
29                 CAN                188.20
128                NLD                186.84
100                LIE                186.36
132                NZL                178.80
75                 HUN                174.95

In [218]:
# Sort the DataFrame by 'Mobile Speed Mbps' in descending order and select the top 25
top_25_mobile_speed = speedtest_index_df.sort_values(by='Mobile Speed Mbps', ascending=False).head(25)

# Print the top 25 countries
print("Top 25 Countries Based on Mobile Speed (Mbps):")
print(top_25_mobile_speed[['ISO 3166-1 alpha-3', 'Mobile Speed Mbps']])

Top 25 Countries Based on Mobile Speed (Mbps):
    ISO 3166-1 alpha-3  Mobile Speed Mbps
4                  ARE             398.51
144                QAT             344.34
94                 KWT             239.83
93                 KOR             141.23
128                NLD             133.44
47                 DNK             130.05
129                NOR             128.77
148                SAU             122.28
16                 BGR             117.64
104                LUX             114.42
151                SGP             114.30
17                 BHR             113.07
32                 CHN             108.37
123                MYS             106.94
182                USA             103.26
25                 BRN             103.08
58                 FRA             103.08
8                  AUS              97.86
163                SWE              97.24
77                 IND              96.38
113                MKD              94.33
30                 CHE       

In [220]:
# Filter the DataFrame for the region "South Asia"
south_asia_df = speedtest_index_df[speedtest_index_df['Region'] == 'South Asia']

# Select the relevant columns: country (ISO code), Mobile Speed, and Fixed Broadband Speed
south_asia_speeds = south_asia_df[['ISO 3166-1 alpha-3', 'Mobile Speed Mbps', 'Fixed Broadband Mbps']]

# Print the results
print("Countries in South Asia with Mobile and Fixed Broadband Speeds:")
print(south_asia_speeds)

Countries in South Asia with Mobile and Fixed Broadband Speeds:
    ISO 3166-1 alpha-3  Mobile Speed Mbps  Fixed Broadband Mbps
0                  AFG               8.48                  3.63
15                 BGD              27.76                 48.14
26                 BTN                NaN                   NaN
77                 IND              96.38                 63.79
101                LKA              18.96                 22.94
110                MDV                NaN                 13.27
130                NPL                NaN                 71.62
134                PAK              19.59                 15.52


In [ ]:
#Can you write a few lines about the state of the internet, technology, and data? Reflect on where
# we stand in this relentless race.

#Ans: Well, based on these results, we stand at a very dismal stage, where even Bangladesh is ahead of us in terms of speed. But when we look at the
#cost per mb, Pakistan is one of the top countries havign the cheapest internet.  

Question 5: Please refer to the COLDAT Colonies CSV File. This data is sourced from [Becker, 2019] and has
been simplified to include indicator variables for major European colonizers: Belgium, Italy, Germany,
France, Britain, the Netherlands, Spain, and Portugal. Note that a country may have multiple colonizers,
while the column ”last colonizer” indicates the single colonizer from which the country later gained
independence.
• Convert the following columns to categorical type: Belgium, Italy, Germany, France, Britain, the
Netherlands, Spain, Portugal, and ”last colonizer.”
• Determine the number of countries that remain colonies using the ”last colonizer” column.
• Calculate the total number of colonies for each colonizer: Belgium, Italy, Germany, France, Britain,
the Netherlands, Spain, and Portugal.
• Compare the number of colonies listed in the ”last colonizer” column with those indicated by the
individual colonizer columns. If there are differences, identify how many countries fall under both
categories. For example, consider countries that were colonized by both Britain and France.

In [232]:
# Read the COLDAT Colonies CSV file
coladat_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/COLDAT_colonies.csv')

# Verify the conversion by checking the data types
print(coladat_df.dtypes)

ISO_Alpha3_Code    object
britain             int64
france              int64
spain               int64
germany             int64
portugal            int64
netherlands         int64
belgium             int64
italy               int64
last_colonizer     object
dtype: object


In [230]:
# Convert the 'last_colonizer' column to categorical type
coladat_df['last_colonizer'] = coladat_df['last_colonizer'].astype('category')

# Verify the conversion by checking the data types
print(coladat_df['last_colonizer'].dtype)

category


In [234]:
# Count the occurrences of each last colonizer
colonizer_counts = coladat_df['last_colonizer'].value_counts()

# Display the number of countries that remain colonies for each colonizer
print("Number of countries that remain colonies by last colonizer:")
print(colonizer_counts)

Number of countries that remain colonies by last colonizer:
last_colonizer
britain        65
france         26
spain          22
portugal        7
belgium         3
netherlands     2
germany         2
Name: count, dtype: int64


In [236]:
# Specify the list of colonizers
colonizers = [
    'belgium', 
    'italy', 
    'germany', 
    'france', 
    'britain', 
    'netherlands', 
    'spain', 
    'portugal'
]

# Filter the DataFrame for the specified colonizers
colonizer_counts = coladat_df[coladat_df['last_colonizer'].isin(colonizers)]

# Count the occurrences of each colonizer
total_colonies = colonizer_counts['last_colonizer'].value_counts()

# Display the total number of colonies for each colonizer
print("Total number of colonies for each colonizer:")
print(total_colonies)

Total number of colonies for each colonizer:
last_colonizer
britain        65
france         26
spain          22
portugal        7
belgium         3
netherlands     2
germany         2
Name: count, dtype: int64


Question 6
Refer to these CSV files: 1) Africa, 2) Americas, 3) Asia, 4) Europe, 5) Australia Oceania, 6)
Transcontinental States . These files contain country names, the date of sovereignty acquisition, and
descriptions of how sovereignty was acquired. For more details, visit the Wikipedia page. Each of these
data frames has three same columns 1) Country Name, 2) Sovereignty Acquisition Date, 3) Sovereignty
Acquisition Description.
• Read the CSV files and assign corresponding name such as: Africa df, Americas df, Asia df,
Australia Oceania df, Europe df, and Transcontinental states df
• Create a column -“continent name” in each of these 6 files, fill with appropriate name of the content.
• Combine all six DataFrames into a single unified DataFrame and name it continents df.
• Ensure that the resulting DataFrame retains all relevant columns and number of rows from each
continent.
• Clean the Data: Remove any unnecessary notes, references, or sources attached to the values in
any column.
• Extract valid four-digit years starting with ”15”, ”16”, ”17”, ”18”, ”19”, or ”20” from the Sovereignty
Acquisition Date column and store them in a new column named Year.
• Fill any missing values in the Year column with the default value of 1400 and convert the column
to a numeric data type, ensuring that any conversion errors result in NaN.
• This default is used because some countries may have acquisition dates that are earlier or in BC
(Before Christ), as indicated by entries such as ”600 BC”, ”218 BC”, and ”3 September 301”.
The value 1400 serves as a placeholder for these cases where specific years cannot be accurately
extracted.
• Extract colonizer: Belgium, Italy, Germany, France, Britain, the Netherlands, Spain, Portugal,
other using the column Sovereignty Acquisition Description.
• A country may have multiple entries for different dates of sovereignty acquisition, Therefore remove
duplicate entries based on Country Name, ensuring that you keep the row with the latest Year
sovereignty acquisition for each country.
• Convert the colonizer columns to categorical type. Compare the number of colonies for each
colonizer using this data alongside the COLDAT dataset. Don’t worry if the numbers don’t match
exactly; the goal is to achieve a relatively comparable analysis.comparable analysis.

In [240]:
# Read the CSV files and assign them to DataFrames
africa_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Africa.csv')
americas_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Americas.csv')
asia_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Asia.csv')
europe_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Europe.csv')
australia_oceania_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Australia_Oceania.csv')  
transcontinental_states_df = pd.read_csv('C:/Users/uy128/Desktop/Academics/IBA/Data Analytics for Economists/DataSets/Transcontinental_states.csv')

# Print the first few rows of each DataFrame to verify the contents
print("Africa DataFrame:")
print(africa_df.head())
print("\nAmericas DataFrame:")
print(americas_df.head())
print("\nAsia DataFrame:")
print(asia_df.head())
print("\nEurope DataFrame:")
print(europe_df.head())
print("\nAustralia Oceania DataFrame:")
print(australia_oceania_df.head())
print("\nTranscontinental States DataFrame:")
print(transcontinental_states_df.head())

Africa DataFrame:
  Country_Name Sovereignty_Acquisition_Date  \
0      Algeria                       202 BC   
1      Algeria                  5 July 1962   
2       Angola             11 November 1975   
3        Benin                1 August 1960   
4     Botswana            30 September 1966   

                 Sovereignty_Acquisition_Description  
0  Massinissa unifies Numidia which extended from...  
1  Algeria reclaims sovereignty over the entire A...  
2                         Independence from Portugal  
3                           Independence from France  
4               Independence from the United Kingdom  

Americas DataFrame:
          Country_Name Sovereignty_Acquisition_Date  \
0  Antigua and Barbuda              1 November 1981   
1            Argentina                  25 May 1810   
2            Argentina                  9 July 1816   
3              Bahamas               7 January 1964   
4              Bahamas                 10 July 1973   

                 

In [242]:
# Add "continent name" column to each DataFrame
africa_df['continent name'] = 'Africa'
americas_df['continent name'] = 'Americas'
asia_df['continent name'] = 'Asia'
europe_df['continent name'] = 'Europe'
australia_oceania_df['continent name'] = 'Australia Oceania'
transcontinental_states_df['continent name'] = 'Transcontinental States'

# Combine all DataFrames into a single DataFrame
continents_df = pd.concat([africa_df, americas_df, asia_df, europe_df, australia_oceania_df, transcontinental_states_df], ignore_index=True)

# Display the combined DataFrame
print("Combined DataFrame (continents_df):")
print(continents_df.head())  # Display the first few rows to verify
print("\nTotal number of rows in continents_df:", len(continents_df))

Combined DataFrame (continents_df):
  Country_Name Sovereignty_Acquisition_Date  \
0      Algeria                       202 BC   
1      Algeria                  5 July 1962   
2       Angola             11 November 1975   
3        Benin                1 August 1960   
4     Botswana            30 September 1966   

                 Sovereignty_Acquisition_Description continent name  
0  Massinissa unifies Numidia which extended from...         Africa  
1  Algeria reclaims sovereignty over the entire A...         Africa  
2                         Independence from Portugal         Africa  
3                           Independence from France         Africa  
4               Independence from the United Kingdom         Africa  

Total number of rows in continents_df: 367


In [306]:
# Function to clean the DataFrame
def clean_data(df):
    # Example cleaning: Remove any text in parentheses and trim whitespace
    for column in df.columns:
        # Remove text in parentheses and trim whitespace
        df.loc[:, column] = df[column].astype(str).str.replace(r'\s*\(.*?\)', '', regex=True).str.strip()
    return df

# Clean the combined DataFrame
continents_df = clean_data(continents_df)

# Optionally, print the cleaned DataFrame to verify
print("Cleaned Combined DataFrame (continents_df):")
print(continents_df.head())

Cleaned Combined DataFrame (continents_df):
  Country_Name Sovereignty_Acquisition_Date  \
0  Afghanistan                         1747   
1      Albania             28 November 1944   
2      Algeria                  5 July 1962   
3      Andorra                         1278   
4       Angola             11 November 1975   

                 Sovereignty_Acquisition_Description continent name  Year  \
0                                     Durrani Empire           Asia  1747   
1  Albanian state re-established after Italian/Ge...         Europe  1944   
2  Algeria reclaims sovereignty over the entire A...         Africa  1962   
3                           Independence from Aragon         Europe  1400   
4                         Independence from Portugal         Africa  1975   

  Colonizer  
0     other  
1     other  
2    france  
3     other  
4  portugal  


In [256]:
# Function to extract valid years
def extract_years(df):
    # Correct regex to find four-digit years starting with 15, 16, 17, 18, 19, or 20
    df['Year'] = df['Sovereignty-Acquisition Date'].str.extract(r'(?<!\d)([1][5-9]\d{2}|[2][0]\d{2})(?!\d)')[0]
    return df

# Extract years and add to the DataFrame
continents_df = extract_years(continents_df)

# Optionally, print the updated DataFrame to verify
print("Updated Combined DataFrame with Year column:")
print(continents_df[['Sovereignty Acquisition Date', 'Year']].head())

Updated Combined DataFrame with Year column:
  Sovereignty_Acquisition_Date  Year
0                       202 BC   NaN
1                  5 July 1962  1962
2             11 November 1975  1975
3                1 August 1960  1960
4            30 September 1966  1966


In [264]:
# Fill missing values in the 'Year' column with 1400 and assign back to the column
continents_df['Year'] = continents_df['Year'].fillna(1400)

# Convert the 'Year' column to numeric, setting errors='coerce' to handle conversion errors
continents_df['Year'] = pd.to_numeric(continents_df['Year'], errors='coerce')

# Optionally, print the updated DataFrame to verify
print("Updated Combined DataFrame with Year column:")
print(continents_df[['Sovereignty_Acquisition_Date', 'Year']].head())

Updated Combined DataFrame with Year column:
  Sovereignty_Acquisition_Date  Year
0                       202 BC  1400
1                  5 July 1962  1962
2             11 November 1975  1975
3                1 August 1960  1960
4            30 September 1966  1966


In [266]:
print(continents_df.head())

  Country_Name Sovereignty_Acquisition_Date  \
0      Algeria                       202 BC   
1      Algeria                  5 July 1962   
2       Angola             11 November 1975   
3        Benin                1 August 1960   
4     Botswana            30 September 1966   

                 Sovereignty_Acquisition_Description continent name  Year  
0  Massinissa unifies Numidia which extended from...         Africa  1400  
1  Algeria reclaims sovereignty over the entire A...         Africa  1962  
2                         Independence from Portugal         Africa  1975  
3                           Independence from France         Africa  1960  
4               Independence from the United Kingdom         Africa  1966  


In [268]:
# Function to extract colonizer
def extract_colonizer(description):
    description = description.lower()  # Convert to lowercase for case-insensitive matching
    for colonizer in colonizers:
        if colonizer in description:
            return colonizer.capitalize()  # Capitalize the first letter for consistency
    return 'Other'

# Apply the function to create a new column 'Colonizer'
continents_df['Colonizer'] = continents_df['Sovereignty_Acquisition_Description'].apply(extract_colonizer)

# Optionally, print the updated DataFrame to verify
print("Updated Combined DataFrame with Colonizer column:")
print(continents_df[['Sovereignty_Acquisition_Description', 'Colonizer']].head())

Updated Combined DataFrame with Colonizer column:
                 Sovereignty_Acquisition_Description Colonizer
0  Massinissa unifies Numidia which extended from...     Other
1  Algeria reclaims sovereignty over the entire A...    France
2                         Independence from Portugal  Portugal
3                           Independence from France    France
4               Independence from the United Kingdom     Other


In [270]:
continents_df['Colonizer'].value_counts()

Colonizer
Other          270
Spain           43
France          26
Portugal        11
Britain          7
Netherlands      4
Belgium          3
Germany          3
Name: count, dtype: int64

In [274]:
# Sort the DataFrame by 'Country Name' and 'Year' in descending order
continents_df.sort_values(by=['Country_Name', 'Year'], ascending=[True, False], inplace=True)

# Drop duplicates, keeping the first occurrence (which will be the latest due to sorting)
continents_df = continents_df.drop_duplicates(subset='Country_Name', keep='first')

# Optionally, reset the index for better readability
continents_df.reset_index(drop=True, inplace=True)

# Print the updated DataFrame to verify
print("DataFrame after removing duplicates:")
print(continents_df.head())

DataFrame after removing duplicates:
  Country_Name Sovereignty_Acquisition_Date  \
0  Afghanistan                         1747   
1      Albania             28 November 1944   
2      Algeria                  5 July 1962   
3      Andorra                         1278   
4       Angola             11 November 1975   

                 Sovereignty_Acquisition_Description continent name  Year  \
0                                     Durrani Empire           Asia  1747   
1  Albanian state re-established after Italian/Ge...         Europe  1944   
2  Algeria reclaims sovereignty over the entire A...         Africa  1962   
3                           Independence from Aragon         Europe  1400   
4                         Independence from Portugal         Africa  1975   

  Colonizer  
0     Other  
1     Other  
2    France  
3     Other  
4  Portugal  


In [298]:

# Normalize the 'Colonizer' column in continents_df to lowercase
continents_df.loc[:, 'Colonizer'] = continents_df['Colonizer'].str.lower()

# Normalize the 'last_colonizer' column in coladat_df to lowercase
coladat_df['last_colonizer'] = coladat_df['last_colonizer'].str.lower()

# Count the number of colonies for each colonizer in continents_df
colonizer_counts = continents_df['Colonizer'].value_counts()

# Count the number of colonies for each last_colonizer in COLDAT
coladat_counts = coladat_df['last_colonizer'].value_counts()

# Create a comparison DataFrame
comparison_df = pd.DataFrame({
    'Continents Data Count': colonizer_counts,
    'COLDAT Data Count': coladat_counts
}).fillna(0)  # Fill NaN with 0 for easier comparison

# Convert counts to integer type
comparison_df = comparison_df.astype(int)

# Print the comparison DataFrame
print("Comparison of the number of colonies for each colonizer:")
print(comparison_df)

Comparison of the number of colonies for each colonizer:
             Continents Data Count  COLDAT Data Count
belgium                          3                  3
britain                          2                 65
france                          21                 26
germany                          2                  2
netherlands                      3                  2
other                          143                  0
portugal                         6                  7
spain                           18                 22
